Add realm owner feature.

sahil839 opened the pull request:
<!-- What's this PR for?  (Just a link to an issue is fine.) -->
This PR has backend code for realm owner feature.
 

- The user who creates the realm is made realm owner.
- Migration is written to make oldest realm admin as realm owner.
- Add or remove realm owner status of user.
- Last realm owner cannot be removed from realm owner status and also cannot be deactivated from realm.

**Testing Plan:** <!-- How have you tested? -->


**GIFs or Screenshots:** <!-- If a UI change.  See:
  https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
  -->


<!-- Also be sure to make clear, coherent commits:
  https://zulip.readthedocs.io/en/latest/contributing/version-control.html
  -->

sahil839 commented on zerver/views/users.py:
```diff
@@ -91,17 +92,32 @@ def update_user_backend(request: HttpRequest, user_profile: UserProfile, user_id
+
+    if (((is_owner is None and target.is_realm_owner) or is_owner) and
+            ((is_admin is None and target.is_realm_admin and not target.is_realm_owner) or is_admin)):
+        return json_error(_("Organization administrators cannot be organization owners"))
```
This checks ensures two things-
- More than one value cannot be true.
- Existing role should also be set to false along with the new role passed as true.

sahil839 commented on zerver/views/users.py:
```diff
@@ -91,17 +92,32 @@ def update_user_backend(request: HttpRequest, user_profile: UserProfile, user_id
             role = UserProfile.ROLE_REALM_ADMINISTRATOR
+        if role is None and not target.is_realm_owner:
+            role = UserProfile.ROLE_MEMBER
 
```
Conditions here are bit different and complicated because `is_realm_admin` is true for both admins and owners.

timabbott replied:
I think we should rework how this function works, as the new version is unmanageable.  I'd like us to create a `def parse_requested_role(is_owner, is_admin, is_guest)` that just looks at those properties to determine what role was requested (if any of them is not None) and returns that.  And then we can do a single conditional comparing it with `user_profile.role` to determine whether we need to change the role.

Once we've isolated that, we can open a separate follow-up issue for allowing accepting a `role` parameter in the API, with that `parse_requested_role` function present for backwards-compatibility.

We should do this as a new first commit with just the `is_admin` and `is_guest` parameters, that we can merge happily, and then a small tweak on top of that to pass the new `is_owner` option.

sahil839 replied:
Where we should add checks for ensuring that existing role is passed False? (Checking those in this new function will also need target to be passed in it.)

sahil839 commented on zerver/models.py:
```diff
@@ -1116,6 +1123,10 @@ def is_realm_admin(self, value: bool) -> None:
+    def is_realm_owner(self) -> bool:
+        return self.role == UserProfile.ROLE_REALM_OWNER
+
     @property
```
I was not sure whether we want to add setter for this as it was not required till now.

timabbott replied:
Yeah, we can probably skip the setter for this.

sahil839 commented:
Forgot to add support for inviting user as owner, will add it soon.

timabbott commented on zerver/openapi/zulip.yaml:
```diff
@@ -1363,6 +1367,14 @@ paths:
+      - name: is_owner
+        in: query
+        description: |
+          Whether the target user is an owner.
```
Let's consistently say "target user is an organization owner."  -- i.e. always put the "organization" word in there.

timabbott commented on zilencer/management/commands/populate_db.py:
```diff
@@ -283,7 +283,8 @@ def handle(self, **options: Any) -> None:
             desdemona = get_user_by_delivery_email("desdemona@zulip.com", zulip_realm)
-            do_change_user_role(desdemona, UserProfile.ROLE_REALM_ADMINISTRATOR)
+
+            do_change_user_role(desdemona, UserProfile.ROLE_REALM_OWNER)
```
Remove the extra space.  But I agree with the decision to tag Iago as a realm administrator (but not owner); that limits how much we need to change tests.

timabbott commented on zerver/tests/test_audit_log.py:
```diff
@@ -57,6 +57,8 @@ def test_change_role(self) -> None:
         do_change_user_role(user_profile, UserProfile.ROLE_GUEST)
         do_change_user_role(user_profile, UserProfile.ROLE_MEMBER)
+        do_change_user_role(user_profile, UserProfile.ROLE_REALM_OWNER)
+        do_change_user_role(user_profile, UserProfile.ROLE_MEMBER)
```
This change should have also modified the check function.  I went ahead and added this and then merged this commit for simplicity; see 7d9d2414ca631bb6e31b369d03156e3a82cd8112.

timabbott commented:
I refactored out a few of the prep commits so that they have the coverage needed to pass tests and merged them.  I think we should be aiming to merge this branch incrementally with the migration code near the end; we can put in place most of the infrastructure with good tests before then with it just being impossible to create an owner user in production.

timabbott commented on zerver/tests/test_signup.py:
```diff
@@ -3899,7 +3899,6 @@ def test_do_not_deactivate_final_admin(self) -> None:
 
     def test_do_not_deactivate_final_user(self) -> None:
         realm = get_realm('zulip')
-        do_change_user_role(self.example_user("desdemona"), UserProfile.ROLE_MEMBER)
```
Why did you remove this?  I think the actual change we want to make is that we should have a prep commit replacing any queries filtering on `role=UserProfile.ROLE_REALM_ADMINISTRATOR` to use a `__in` clause in the few places that appears in the codebase.

sahil839 replied:
This was added previously so that we do not have to remove `Desdemona` separately as we wanted to come to a stage with last user in realm.

I actually intended to change the query to exclude realm owner only and keep realm owner as the last user to check deactivating of final user and I missed to change that.

timabbott replied:
Yeah, I think ultimately we want this test to confirm that the last owner can't be removed.

sahil839 replied:
To be clear, this test is to confirm that last user cannot be deactivated, the test for last owner is separte.

timabbott commented on zerver/tests/test_users.py:
```diff
@@ -120,20 +120,21 @@ def test_updating_non_existent_user(self) -> None:
 
     def test_admin_api(self) -> None:
-        self.login('desdemona')
+        self.login('hamlet')
```
I don't understand the change here; why can't we do all these operations as desdemona, the realm owner user?  

That's certainly the intent of how we wanted this test to work; I don't see why we'd want to make hamlet an admin instead.

sahil839 replied:
As desdemona was changed to owner in this commit and we check `is_admin` flag for the requesting user, that is the reason why I changed it to hamlet to  correctly check is_admin flag.

timabbott replied:
But shouldn't `.is_admin` be True for desdemona following other changes we've made?

sahil839 replied:
But the change of `is_admin as True for owners` is in the next commit. And the next commit requires an owner for tests, so I have ordered commit in this way.

Moreover in further commits, as we have removed the restriction on removing last admin, we just require one admin in this test, so we do not require hamelt as admin anyway at the end.

timabbott commented:
Thanks for your work on this @sahil839!  I think this is promising, and we have a few more big chunks we should do as preparatory refactors, but after that we'll be pretty close to being able to integrate this.

Have you looked at the UI code for displaying users who are realm owners correctly?

sahil839 commented:
I have left some xomments above.
Regarding UI, I have not read code thoroughly, but I have a rough idea of where and how to make changes.
I am starting with UI changes side by side today.

timabbott commented:
I think this branch will be cleaned up substantially when rebased on top of the `role` dropdown PR :)

sahil839 commented:
I have made some changes here to backend. Will probably be able to complete the frotnend work by tomorrow.

sahil839 commented:
I wanted to make sure whether I need to change the changelog.md file in for feature level 7 or add a feature level.
It would be helpful to know when we add new feature level in this file.

timabbott commented:
Generally you should always add a new feature level whenever making any changes to the client/server API; they're cheap.  I'll post a reminder on this in #general.

timabbott commented:
I think the next change we should extract and try to integrate, maybe via a separate small PR, are these commits:
* `populate_db: Change 'Desdemona` to realm owner.`
* `users: Add 'is_owner' field to user dict.`

Both of those commits should be mergeable without changing any user-facing behavior, and they represent most of the remaining complexity that could be pulled out of this PR.

(Please make sure each passes CI individually).

timabbott commented:
Can you rebase this now that we've merged the next batch of prep PRs?

The big thing I'd like you to think through carefully for the remainder of this PR is the tests; this PR adds a lot of test code, and I'm not sure it's organized very well.

timabbott commented:
@sahil839 I think for the first commit, we probably will want to do the same sort of migration to pass `role` into `do_create_user`, at least, as a separate prep PR.

I was going to comment on API changes in `create_user_backend`, but apparently we don't have `role` as a supported argument there at all, which is definitely an oversight.  Probably worth adding it in that prep PR.

timabbott commented:
Let me know when you have this rebased again; I think we're pretty close to having extracted and merged everything that could be.

sahil839 commented:
@timabbott I have rebased the PR and have also added commit for changes in events.py.
(Have not changed the mogration yet as discussed in czo, will do it soon).

sahil839 commented:
@timabbott please take a look.
